src/context/WebsocketContext.jsx

```diff
@@ -0,0 +1,159 @@
+import React, { createContext, useEffect, useState, useContext } from 'react'
+
+import { useCurrentWallet } from './WalletContext'
+
+const WEBSOCKET_RECONNECT_DELAY_STEP = 1_000
+const WEBSOCKET_RECONNECT_MAX_DELAY = 10_000
+
+// webservers will close a websocket connection on inactivity (e.g nginx default is 60s)
+// specify the time in milliseconds at least one 'keepalive' message is sent
+const WEBSOCKET_KEEPALIVE_MESSAGE_INTERVAL = 30_000
+
+// return delay in milliseconds to attempt reconnecting after the connection has been lost
+const connectionRetryDelayLinear = (attempt = 0) => {
+  // linear increase per attempt by `step` amount till `max` is reached
+  const delay = Math.max(WEBSOCKET_RECONNECT_DELAY_STEP, WEBSOCKET_RECONNECT_DELAY_STEP * attempt)
+  return Math.min(delay, WEBSOCKET_RECONNECT_MAX_DELAY)
+}
+
+// path that will be proxied to the backend server
+const WEBSOCKET_ENDPOINT_PATH = '/jmws'
+
+// possible values for property `coinjoin_state` in websocket messages
+const CJ_STATE_TAKER_RUNNING = 0
+const CJ_STATE_MAKER_RUNNING = 1
+const CJ_STATE_NONE_RUNNING = 2
+
+const createWebSocket = () => {
+  const { protocol, host } = window.location
+  const scheme = protocol === 'https:' ? 'wss' : 'ws'
+  const websocket = new WebSocket(`${scheme}://${host}${WEBSOCKET_ENDPOINT_PATH}`)
+
+  if (process.env.NODE_ENV !== 'production') {
+    websocket.onopen = () => {
+      console.debug('websocket connection openend')
+    }
+
+    websocket.onclose = () => {
+      console.debug('websocket connection closed')
+    }
+
+    websocket.onmessage = (event) => {
+      const data = JSON.parse(event?.data)
+      console.debug('websocket message', data)
+    }
+  }
+
+  websocket.onerror = (error) => {
+    console.error('websocket error', error)
+  }
+
+  return websocket
+}
+
+const initialWebsocket = createWebSocket()
+
+const WebsocketContext = createContext()
+
+/**
+ * Provider of a websocket connection to jmwalletd.
+ *
+ * See Websocket docs: https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/v0.9.5/docs/JSON-RPC-API-using-jmwalletd.md#websocket
+ */
+const WebsocketProvider = ({ children }) => {
+  const [websocket, setWebsocket] = useState(initialWebsocket)
+  const [websocketState, setWebsocketState] = useState(initialWebsocket.readyState)
+  const setConnectionErrorCount = useState(0)[1]
+  const currentWallet = useCurrentWallet()
+
+  // update websocket state based on open/close events
+  useEffect(() => {
+    const onStateChange = () => setWebsocketState(websocket.readyState)
+
+    websocket.addEventListener('open', onStateChange)
+    websocket.addEventListener('close', onStateChange)
+
+    return () => {
+      websocket && websocket.removeEventListener('close', onStateChange)
+      websocket && websocket.removeEventListener('open', onStateChange)
+    }
+  }, [websocket])
```

Comment:
Is `websocketState` ever different from using `websocket.readyState` directly?

Reply:
Yes, small differences exist. `websocketState` will only once be `connecting`, subsequently only changes between `open` and `closed`. It would be great to have a way to react on state changes without registering event listeners in every component that wants to act on them. Do you know how that can be achieved?

Reply:
No not really. Let's keep it like this then! 👍